File: classe/exercicio1.cpp
#include <iostream>
#include <iomanip>
#include <stdexcept>
using namespace std;

class Tipo {
    private:
        int codigo;
        float percentual;
    public:
        Tipo () { setCodigo(1); }
        void setCodigo(int num) {
            try {
                if (num == 1) {
                    codigo = num;
                    setPercentual(10);
                } else if (num == 2) {
                    codigo = num;
                    setPercentual(20);
                }
                else throw invalid_argument ("Codigo invalido!");
            } catch (exception &e) { cerr << e.what() << endl; }
        }
        int getCodigo() { return codigo; }
        string formataCod () {
            if (codigo == 1) return "Alimentacao";
            else if (codigo == 2) return "Limpeza";
            else return "Desconhecido";
        }
        float getPercentual() { return percentual; }
        void preenche () {
            int n;
            cout << "Tipo do produto: ";
            cin >> n; setCodigo(n);
            cout << endl;
        }
        void exibe () { cout << "Tipo do produto: " << formataCod() << endl << "Percentual de imposto: " << getPercentual() << endl; }
    private:
        void setPercentual(float val) {
            try {
                if (val >= 0) percentual = val;
                else throw invalid_argument ("Codigo invalido!");
            } catch (exception &e) { cerr << e.what() << endl; }
        }
};

class Produto {
    private:
        string descricao;
        Tipo tipo;
        float preco, imposto;
    public:
        Produto() {
            setDescricao("nenhum");
            tipo.setCodigo(1);
            setPreco(1);
            setImposto();
        }
        Produto(string d, int n, float p) {
            setDescricao(d);
            tipo.setCodigo(n);
            setPreco(p);
            setImposto();
        }
        void setDescricao(string d) {
            try {
                if (d.length() > 3) descricao = d;
                else throw invalid_argument ("Descricao invalida!");
            } catch (exception &e) { cerr << e.what() << endl; }
        }
        string getDescricao() { return descricao; }
        void setPreco(float p) {
            try {
                if (p > 0) preco = p;
                else throw invalid_argument ("Preco invalido!");
            } catch (exception &e) { cerr << e.what() << endl; }
        }
        float getPreco() { return preco; }
        float getImposto() { return imposto; }
        float getFinal() { return getPreco()+getImposto(); }
        void preenche() {
            string txt;
            float val;
            cout << "Descricao do produto: ";
            getline(cin, txt); setDescricao(txt);
            cout << "Preco do produto: ";
            cin >> val; setPreco(val);
            setImposto();
            tipo.preenche();
        }
        void exibe() {
            cout << "Descricao: " << getDescricao() << endl << "Preco: " << getPreco() << " reais" << endl << "Imposto: " << getImposto() << " reais" << endl;
            tipo.exibe();
        }
    private:
        void setImposto() { imposto = getPreco()*tipo.getPercentual()/100; }
};

int main () {
    Produto p;
    p.preenche();
    p.exibe();
    cout << "Preco final do produto: " << p.getFinal() << endl;
}

File: AEDS2-C&JAVA/classe/exercicio1.java
import java.util.Scanner;

class Tipo {
    private int codigo;
    private double percentual;
    public Tipo () { setCodigo(1); }
    public void setCodigo(int num) {
        try {
            if (num == 1) {
                codigo = 1; percentual = 10;
            } else if (num == 2) {
                codigo = 2; percentual = 20;
            } else throw new IllegalArgumentException("Código inválido!");
        } catch (Exception e) { System.err.println(e.getMessage()); }
    }
    public int getCodigo() { return codigo; }
    public string formataCod() {
        if (codigo == 1) return "Alimentação";
        else if (codigo == 2) return "Transporte";
        else return "Desconhecido";
    }
    public double getPercentual() { return percentual; }
    public void preenche(Scanner sc) {
        int n;
        Systema.out.print("Tipo do produto: ");
        n = sc.nextInt();
        setCodigo(n);
        System.out.println();
    }
    public void exibe() {
        System.out.println("Tipo do produto: " + formataCod());
        System.out.println("Percentual de imposto: " + getPercentual() + "%");
    }
}

class Produto {
    private String descricao;
    private Tipo tipo = new Tipo();
    private double preco, imposto;

    public Produto() {
        setDescricao("nenhum");
        tipo.setCodigo(1);
        setPreco(1);
        setImposto();
    }
    public Produto(String d, int n, double p) {
        setDescricao(d);
        tipo.setCodigo(n);
        setPreco(p);
        setImposto();
    }
    public void setDescricao(String d) {
        try {
            if (d.length() > 0) descricao = d;
            else throw new IllegalArgumentExcpetion("Descrição inválida!");
        } catch (Exception e) { System.err.println(e.getMessage()); }
    }
    public String getDescricao() { return descricao; }
    public void setPreco(double p) {
        try {
            if(p > 0) preco = p;
            else throw new IllegalArgumentException("Preço inválido!");
        } catch (Exception e) { System.err.println(e.getMessage()); }
    }
    public double getPreco() { return preco; }
    public double getImposto() { return imposto; }
    public double getFinal() { return getPreco() + getImposto(); }
    public void preenche(Scanner sc) {
        String txt;
        double val;
        System.out.print("Descrição do produto: ");
        txt = sc.nextLine(); setDescricao(txt);
        System.out.print("Preço do produto: ");
        val = sc.nextdouble(); setPreco(val);
        setImposto();
        tipo.preenche(sc);
    }
    public void exibe() {
        System.out.println("Descrição: " + getDescricao());
        System.out.println("Preço: " + getPreco() + " reais");
        System.out.println("Imposto: " + getImposto() + " reais");
        tipo.exibe();
    }
    private void setImposto() { imposto = getPreco() * tipo.getPercentual()/100;  }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Produto p = new Produto();
        p.preenche(sc); p.exibe();
        System.out.println("Preço final do produto: " + p.getFinal() + " reais");
    }
}
